src/uu/tail/src/tail.rs

```diff
@@ -125,127 +187,302 @@ impl Settings {
+                        ""
+                    };
+                    show_error!(
+                        "{}: cannot follow end of this type of file{}",
+                        display_name.display(),
+                        msg
+                    );
+                }
+                if !(settings.follow == Some(FollowMode::Name) && settings.retry) {
+                    // skip directory if not retry
+                    continue;
+                }
+            } else {
+                // TODO: [2021-10; jhscheer] how to handle block device or socket?
+                todo!();
             }
-        } else {
-            let path = Path::new(filename);
-            if path.is_dir() {
-                continue;
+        }
+
+        let metadata = path.metadata().ok();
+
+        if display_name.is_stdin() && path_is_tailable {
+            if settings.verbose {
+                files.print_header(Path::new(text::STDIN_HEADER), !first_header);
+                first_header = false;
             }
-            let mut file = File::open(&path)
-                .map_err_context(|| format!("cannot open {} for reading", filename.quote()))?;
-            let md = file.metadata().unwrap();
-            if is_seekable(&mut file) && get_block_size(&md) > 0 {
-                bounded_tail(&mut file, &settings.mode, settings.beginning);
-                if settings.follow {
-                    let reader = BufReader::new(file);
-                    readers.push((Box::new(reader), filename));
+
+            let mut reader = BufReader::new(stdin());
+            if !stdin_is_bad_fd() {
+                unbounded_tail(&mut reader, &settings)?;
+                if settings.follow == Some(FollowMode::Descriptor) {
+                    // Insert `stdin` into `files.map`
+                    files.insert(
+                        path.to_path_buf(),
+                        PathData {
+                            reader: Some(Box::new(reader)),
+                            metadata: None,
+                            display_name: PathBuf::from(text::STDIN_HEADER),
+                        },
+                    );
                 }
             } else {
-                let mut reader = BufReader::new(file);
-                unbounded_tail(&mut reader, settings)?;
-                if settings.follow {
-                    readers.push((Box::new(reader), filename));
+                settings.exit_code = 1;
+                show_error!(
+                    "cannot fstat {}: {}",
+                    text::STDIN_HEADER.quote(),
+                    text::BAD_FD
+                );
+                if settings.follow.is_some() {
+                    show_error!(
+                        "error reading {}: {}",
+                        text::STDIN_HEADER.quote(),
+                        text::BAD_FD
+                    );
                 }
             }
+        } else if path_is_tailable {
+            match File::open(&path) {
```

Comment:
this block and the next are pretty long. It is possible to move that into a separate function or too complex?